californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -234,6 +235,13 @@ public Exchange receiveResponse(final Response response, final CorrelationContex
 			return null;
 		} else if (correlationContextMatcher.isResponseRelatedToRequest(exchange.getCorrelationContext(), responseContext)) {
 
+			if (response.getType() == Type.ACK && exchange.getCurrentRequest().getMID() != response.getMID()) {
```

Comment:
Is there a case where we do want to accept responses with the same token but an MID that differs from the request for CON/NON?